File: boj/TwoPointer/BOJ_2473.cpp
/*
 백준 2473 세 용액
 21.08.10
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using dd = double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;

#define INF 2e9
#define endl "\n"

ll minSum = INF * 2;
vector <ll> v;
int n, first, second, third;

int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    freopen("input.txt", "r", stdin);

    cin >> n;

    for (int i = 0; i < n; ++i) {
        ll x; cin >> x;
        v.push_back(x);
    }

    sort(v.begin(), v.end());

    // 용액 하나를 고정시키고 진행
    for (int i = 0; i < n - 2; ++i) {
        int l = i + 1, r = n - 1;

        while (l < r) {
            ll sum = v[i] + v[l] + v[r];

            if (abs(sum) < minSum) {
                minSum = abs(sum);
                first = i, second = l, third = r;
            }

            if (sum > 0) {
                --r;
            } else {
                ++l;
            }
        }
    }

    cout << v[first] << " " << v[second] << " " << v[third] << endl;

    return 0;
}

File: boj/TwoPointer/BOJ_2473.swift
/*
 백준 2473 세 용액
 21.08.10
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let n = Int(readLine()!)!

var arr = readLine()!.split(separator: " ").map{ Int(String($0))! }

arr.sort()

var first = 0, second = 0, third = 0, minSum = Int(4e9)

for i in 0 ..< n - 2 {
  var left = i + 1, right = n - 1;

  while left < right {
    let tempSum = arr[i] + arr[left] + arr[right]

    if abs(tempSum) < minSum {
      minSum = abs(tempSum)
      first = i; second = left; third = right
    }

    if tempSum > 0 {
      right -= 1
    } else {
      left += 1
    }
  }
}

print(arr[first], arr[second], arr[third])